src/engraving/dom/tie.cpp

```diff
@@ -959,20 +1035,48 @@ Note* Tie::endNote() const
     return toNote(endElement());
 }
 
-bool Tie::isConnectingEqualArticulations() const
+bool Tie::isOuterTieOfChord(Grip startOrEnd) const
 {
-    if (!startNote() || !endNote()) {
+    if (m_isInside) {
         return false;
     }
 
-    const Chord* firstChord = startNote()->chord();
-    const Chord* lastChord = endNote()->chord();
+    bool start = startOrEnd == Grip::START;
+    Note* note = start ? startNote() : endNote();
+    if (!note) {
+        return false;
+    }
+
+    Chord* chord = note->chord();
+
+    return (note == chord->upNote() && up()) || (note == chord->downNote() && !up());
+}
 
-    if (!firstChord || !lastChord) {
+bool Tie::hasTiedSecondInside() const
+{
+    Note* note = startNote();
+    if (!note) {
         return false;
     }
 
-    return firstChord->containsEqualArticulations(lastChord)
-           && firstChord->containsEqualTremolo(lastChord);
+    Chord* chord = note->chord();
+    int line = note->line();
+    int secondInsideLine = up() ? line + 1 : line - 1;
+
+    for (Note* otherNote : chord->notes()) {
+        if (otherNote->line() == secondInsideLine && otherNote->tieFor() && otherNote->tieFor()->up() == up()) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool Tie::isCrossStaff() const
+{
+    Note* startN = startNote();
+    Note* endN = endNote();
+
+    return (startN && startN->chord()->staffMove() != 0) || (endN && endN->chord()->staffMove() != 0);
```

Comment:
Is this really the desired formula, or should it be
```c++
    return startN && endN && startN->chord()->vStaffIdx() != endN->chord()->vStaffIdx();
```
or, like Slur::isCrossStaff
```c++
    return startCR() && endCR()
           && (startCR()->staffMove() != 0 || endCR()->staffMove() != 0
               || startCR()->vStaffIdx() != endCR()->vStaffIdx());
```

Reply:
Yes, it's correct as it is (we consider it "cross" even if both endpoints are moved to the same staff). Actually, I think the last OR condition in Slur::isCrossStaff is redundant (it can't be true if at least one of the first two isn't) but I'll verify that another time